src/query/storages/system/src/tables_table.rs

```diff
@@ -124,14 +126,139 @@ where TablesTable<WITH_HISTORY, WITHOUT_VIEW>: HistoryAware
     ) -> Result<DataBlock> {
         let tenant = ctx.get_tenant();
         let catalog_mgr = CatalogManager::instance();
-        let catalogs = catalog_mgr
+        let catalogs: Vec<Arc<dyn Catalog>> = catalog_mgr
             .list_catalogs(&tenant, ctx.session_state())
             .await?
             .into_iter()
             .map(|cat| cat.disable_table_info_refresh())
             .collect::<Result<Vec<_>>>()?;
-        self.get_full_data_from_catalogs(ctx, push_downs, catalogs)
-            .await
+
+        // Check if push_downs only contains projection for 'show tables' column
+        let schema = TablesTable::<WITH_HISTORY, WITHOUT_VIEW>::schema();
+        let mut name_fields_indexes: HashSet<usize> = HashSet::new();
+        for (i, name) in schema.fields.iter().enumerate() {
+            match name.name().as_str() {
+                "name" | "database" | "catalog" | "table_type" => {
+                    name_fields_indexes.insert(i);
+                }
+                _ => {}
+            }
+        }
+        let mut only_show_table_projection = false;
+        let mut catalog_name = String::new();
+        if let Some(push_downs) = &push_downs {
+            if let Some(Projection::Columns(projection_indices)) = &push_downs.projection {
+                // SELECT name AS `Tables_in_tpcds` FROM default.system.tables where database = 'tpcds' and table_type = 'BASE TABLE' and catalog = 'ctl' ORDER BY catalog,database,name
+                if projection_indices.len() == 4 {
+                    only_show_table_projection = projection_indices
+                        .iter()
+                        .any(|field_index| name_fields_indexes.contains(field_index));
+                }
+            }
+            if let Some(filter) = push_downs.filters.as_ref().map(|f| &f.filter) {
+                let expr = filter.as_expr(&BUILTIN_FUNCTIONS);
+                find_eq_filter(&expr, &mut |col_name, scalar| {
+                    if col_name == "catalog" {
+                        if let Scalar::String(catalog) = scalar {
+                            if !catalog_name.contains(catalog) {
+                                catalog_name = catalog.to_string();
+                            }
+                        }
+                    }
+                    Ok(())
+                });
+            }
+        };
+
+        let mut only_iceberg_catalog = false;
```

Comment:
this should not be a special case.

Reply:
Default catalog should check privileges.

Reply:
And we already disable refresh when query system.tables.